inline-r/src/Language/R/Parser.hs

```diff
@@ -0,0 +1,327 @@
+-- each attribute exists, returns list of the attribute name, together
+-- with parser result.
+-- If parser returns @Nothing@ - result is omitted..
+attributes :: Parser s r (Maybe a) -> Parser s r [(String, a)]
+attributes p = do
+   SomeSEXP s <- somesexp
+   let sa = unsafePerformIO $ SomeSEXP <$> R.getAttributes s
+   with sa $ choice
+     [ null *> pure []
+     , do mns <- optional names
+          case mns of
+            Nothing -> return []
+            Just ns -> do
+              ps <- list (length ns) p
+              return $ mapMaybe (\(x,y) -> fmap (x,) y) $ zip ns ps
+     , pure []
+     ]
+
+-- | 'Language.R.Hexp.hexp' lifted to Parser, applies hexp to the current
+-- value and allow to run internal parser on it. Is useful when you need
+-- to inspect data using high level functions from @Language.R@.
+hexp :: SSEXPTYPE ty -> (HExp s ty -> Parser s r a) -> Parser s r a
+hexp ty f = f . H.hexp =<< sexp ty
+
+-- | Returns type of the current SEXP. Can never fail.
+typeOf :: Parser s r R.SEXPTYPE
+typeOf = (\(SomeSEXP s) -> H.typeOf s) <$> somesexp
+
+-- | Return the class of the S3 object, fails otherwise.
+getS3Class :: Parser s r [String]
+getS3Class = charList <$> attribute SString "class"
+--------------------------------------------------------------------------------
+-- Helpers
+--------------------------------------------------------------------------------
+
+-- | Convert String 'SEXP' to the list of 'String's.
+charList :: SEXP s 'R.String -> [String]
+charList (H.hexp -> String v) =
+  map ((\(Char s) -> SV.toString s) . H.hexp) $ SV.toList v
+charList _ = error "Impossible happened."
+
+-- | Get 'dim' attribute.
+dim :: Parser s r [Int]
+dim = go <$> attribute SInt "dim"
+ where
+   go :: SEXP s 'R.Int -> [Int]
+   go (H.hexp -> Int v) = fromIntegral <$> SV.toList v
+   go _ = error "Impossible happened."
+
+-- | Get 'dimnames' attribute.
+dimnames :: Parser s r [[String]]
+dimnames = do
+  s <- attribute SVector "dimnames"
+  case H.hexp s of
+    Vector _ v -> for (SV.toList v) (`with` go)
+  where
+    go = choice [ charList <$> sexp SString
+                , null *> pure []
+                ]
+
+-- | Get 'names' attribute.
+names :: Parser s r [String]
+names = do
+  s <- attribute SString "names"
+  return $ charList s
+
+-- | Get 'rownames' attribute.
+rownames :: Parser s r [String]
+rownames = do s <- attribute SString "row.names"
+              return $ charList s
+
+-- | Execute first parser that will not fail.
+choice :: [Parser r s a] -> Parser r s a
+choice = asum
+
+-- | Parser that is applied to the R @List@ object and returns l
+-- of the results.
+list :: Int -- ^ Number of elements in a list.
+     -> Parser s r (Maybe a) -- ^ Parser to apply to each element
+     -> Parser s r [Maybe a]
+list 0 _ = return []
+list n p = choice
+  [ hexp SList $ \(List car cdr _) -> do
+       v  <- with (SomeSEXP car) p
+       vs <- with (SomeSEXP cdr) $ list (n-1) p
+       return (v:vs)
+  , pure []
+  ]
```

Comment:
hm, perhaps continuations make rewriting the function as I did, unnecessary. Doing `return (v:vs)` after the recursive call does look like it will consume stack space for each recursive call. The list would need to be constructed in an accumulating parameter.